integration/log.go

```diff
@@ -304,70 +303,63 @@ func readbackLogEntries(logID int64, client trillian.TrillianLogClient, params T
+		count := end - index
+		if max := params.ReadBatchSize; count > max {
+			count = max
 		}
 
-		glog.Infof("Reading %d leaves from %d ...", numLeaves, currentLeaf+params.StartLeaf)
-		req := &trillian.GetLeavesByRangeRequest{LogId: logID, StartIndex: currentLeaf + params.StartLeaf, Count: numLeaves}
+		glog.Infof("Reading %d leaves from %d ...", count, index)
+		req := &trillian.GetLeavesByRangeRequest{LogId: logID, StartIndex: index, Count: count}
 		ctx, cancel := getRPCDeadlineContext(params)
 		response, err := client.GetLeavesByRange(ctx, req)
 		cancel()
-
 		if err != nil {
 			return nil, err
 		}
 
-		// Check we got the right leaf count
-		if len(response.Leaves) == 0 {
-			return nil, fmt.Errorf("expected %d leaves log returned none", numLeaves)
+		// Check we got the right number of leaves.
+		if got, want := int64(len(response.Leaves)), count; got != want {
+			return nil, fmt.Errorf("expected %d leaves, got %d", want, got)
 		}
 
-		// Check the leaf contents make sense. Can't rely on exact ordering as queue timestamps will be
-		// close between batches and identical within batches.
-		for l := 0; l < len(response.Leaves); l++ {
-			// Check for duplicate leaf index in response data - should not happen
-			leaf := response.Leaves[l]
-
-			lk := string(leaf.LeafValue)
-			expect[lk]--
+		for _, leaf := range response.Leaves {
+			leaves = append(leaves, leaf)
 
-			if expect[lk] == 0 {
-				delete(expect, lk)
+			// Check that the leaf values are returned in the same counts. Can't rely
+			// on the ordering because they can be inserted in a different order.
+			value := string(leaf.LeafValue)
+			expect[value]--
+			if expect[value] == 0 {
+				delete(expect, value)
 			}
-			leafMap[leaf.LeafIndex] = leaf
 
+			// Check that the MerkleLeafHash field is computed correctly.
 			hash := rfc6962.DefaultHasher.HashLeaf(leaf.LeafValue)
-
-			if got, want := hex.EncodeToString(hash), hex.EncodeToString(leaf.MerkleLeafHash); got != want {
-				return nil, fmt.Errorf("leaf %d hash mismatch expected got: %s want: %s", leaf.LeafIndex, got, want)
+			if got, want := leaf.MerkleLeafHash, hash; !bytes.Equal(got, want) {
+				return nil, fmt.Errorf("leaf %d hash mismatch: got %x want %x", leaf.LeafIndex, got, want)
 			}
 
-			// Ensure that the ExtraData in the leaf made it through the roundtrip. This was set up when
-			// we queued the leaves.
-			if got, want := hex.EncodeToString(leaf.ExtraData), hex.EncodeToString([]byte(strings.Replace(string(leaf.LeafValue), "Leaf", "Extra", 1))); got != want {
-				return nil, fmt.Errorf("leaf %d extra data got: %s, want:%s (%v)", leaf.LeafIndex, got, want, leaf)
+			// Ensure that the ExtraData in the leaf made it through the roundtrip.
+			// This was set up when we queued the leaves.
+			if got, want := leaf.ExtraData, []byte(strings.Replace(value, "Leaf", "Extra", 1)); !bytes.Equal(got, want) {
+				return nil, fmt.Errorf("leaf %d ExtraData: got %x, want %x: %v", leaf.LeafIndex, got, want, leaf)
```

Comment:
This is also pretty weird that this function that otherwise is pretty generic Trillian -> []Leaf has prior knowledge of how the leaves were constructed. This would be better in the outer function, alongside the `expect` checks about what the contents of the leaves are.

Reply:
Done.